cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,98 @@
+                    replicaState -> replicaState.lastCaughtUpTimestamp().isPresent() ? replicaState.lastCaughtUpTimestamp().getAsLong() : -1));
+            boolean canRoll = isQuorumHealthy(info.leaderId(), podId, replicaIdToLastCaughtUpTimestampMap);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Restart pod {} would affect the quorum", podId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart pod {}", podId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the quorum leader id");
+        return quorumInfoFuture.map(info -> info.leaderId()).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    private boolean isQuorumHealthy(int leaderId, int podId,  Map<Integer, Long> replicaStates) {
+        int totalNumOfControllers = replicaStates.size();
+        AtomicInteger numOfCaughtUpControllers = new AtomicInteger();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+        long leaderLastCaughtUpTimestamp = replicaStates.get(leaderId);
+        LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for the leader replica {} is {}", leaderId, leaderLastCaughtUpTimestamp);
+        replicaStates.forEach((replicaId, lastCaughtUpTimestamp) -> {
+            if (lastCaughtUpTimestamp < 0) {
+                LOGGER.warnCr(reconciliation, "No valid lastCaughtUpTimestamp is found for the replica {} ", replicaId);
+            } else {
+                if ((leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp) < controllerQuorumFetchTimeoutMs) {
+                    // skip the controller that we are considering to roll unless it's the leader
+                    if (replicaId != podId || replicaId == leaderId) {
+                        numOfCaughtUpControllers.getAndIncrement();
+                    }
+                    LOGGER.debugCr(reconciliation, "The controller {} has caught up with the quorum leader", replicaId);
+                    LOGGER.debugCr(reconciliation, "The lastCaughtUpTimestamp for controller {} is {}", replicaId, lastCaughtUpTimestamp);
+                } else {
+                    LOGGER.debugCr(reconciliation, "The controller {} is behind from the leader", replicaId);
+                }
+            }
+        });
+        LOGGER.infoCr(reconciliation, "Out of {} voters, there are {} controllers that have caught up with the quorum leader, not including the controller {}", totalNumOfControllers, numOfCaughtUpControllers, podId);
```

Comment:
Is the INFO level really appropriate here? Wouldn't it be printed every reconciliation?

Reply:
I've changed this to debug level.